### PR TITLE
Resolve #90 Integrate bleu_tool.py into texar.evals

### LINKED ISSUE #90: Integrate bleu_tool.py into texar.evals
related issue https://github.com/asyml/texar-pytorch/issues/88

### PR DESCRIPTION
**Pull request by gpengzhi**
#90

### CODE REVIEW

File: docs/code/evals.rst [hunk 1]
```diff
@@ -4,10 +4,15 @@
...
 
-
 BLEU
 ==========
 
+.. spelling::
+    Vaswani
+    al
+    de
+    Tokenize
```
Comment (huzecong) [inline docs/code/evals.rst]: Let's add "Tokenize" to the general word list.

File: texar/evals/bleu_tool.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            ngram = tuple(segment[i:i + order])
+            ngram_counts[ngram] += 1
+    return ngram_counts
+
+
+def compute_bleu(reference_corpus: List[List[str]],
+                 translation_corpus: List[List[str]],
+                 max_order: int = 4,
+                 use_bp: bool = True) -> float:
+    r"""Computes BLEU score of translated segments against references.
```
Comment (ZhitingHu) [inline texar/evals/bleu_tool.py]: the name `compute_bleu` is a bit too vague. Let's call it `corpus_bleu_transformer` (on par w/ `corpus_bleu_moses` that indicates the version of BLEU "moses")

File: texar/evals/bleu_tool.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            for x in range(sys.maxunicode)
+            if unicodedata.category(chr(x)).startswith(prefix)
+        )
+
+
+uregex = UnicodeRegex()
+
+
+def bleu_tokenize(string: str) -> List[str]:
+    r"""Tokenize a string following the official BLEU implementation.
```
Comment (ZhitingHu) [inline texar/evals/bleu_tool.py]: This tokenization follows `mteval-v14.pl`. The standard BLEU implementation is [multi-bleu.perl](https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu.perl). Are the tokenization parts the same?

File: texar/evals/bleu_tool.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    string = uregex.nondigit_punct_re.sub(r"\1 \2 ", string)
+    string = uregex.punct_nondigit_re.sub(r" \1 \2", string)
+    string = uregex.symbol_re.sub(r" \1 ", string)
+    return string.split()
+
+
+def bleu_wrapper(ref_filename: str,
+                 hyp_filename: str,
+                 case_sensitive: bool = False) -> float:
+    r"""Compute BLEU for two files (reference and hypothesis translation).
```
Comment (ZhitingHu) [inline texar/evals/bleu_tool.py]: Let's call it `file_bleu_transformer`, indicating it's for files (on par w/ `sentence_*`, `corpus_*`, etc), and it's the `*_transformer` version.

Even better, can we have a general `file_bleu` with an option for users to choose the BLEU version (if the file reading and preprocessing can be shared)?

Comment (ZhitingHu) [reply texar/evals/bleu_tool.py]: Either way, the docstring should be more detailed, especially indicating which BLEU version is used.

File: texar/evals/bleu_tool.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
+# Copyright 2019 The Texar Authors. All Rights Reserved.
```
Comment (ZhitingHu) [inline texar/evals/bleu_tool.py]: I meant to name the functions as `bleu_tool`. Now that we have other names for the functions, here let's name the file as `bleu_transformer.py`

File: texar/evals/bleu_tool.py (hunk 3 above)
Comment (gpengzhi) [reply texar/evals/bleu_tool.py]: I think `multi-bleu.perl` requires users to tokenize the text before computing BLEU score. It is also mentioned here (https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu.perl#L172). The tokenization (https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/mteval-v14.pl#L926) in `mteval-v14.pl` is pretty standard.

Comment (gpengzhi) [reply texar/evals/bleu_tool.py]: There is [another version](https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu-detok.perl) of `multi-bleu.perl`, which uses the internal tokenization of `mteval-v13a.pl`.

Comment (ZhitingHu) [reply texar/evals/bleu_tool.py]: I see. Let's put this [statement](https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu.perl#L172) (adapt if needed) in our docstring to explain the use of the `bleu_tokenize` function.

**Pull request merged**